### PR TITLE
Video alpha

### PR DESCRIPTION
**Pull request by mattkrick**
I won't have enough time get video to beta this sprint, so i'm merging what i've got behind a feature flag.

This puts video behind a feature flag. you can add yourself by calling `addFeatureFlag` with the `video` param in GraphiQL.

Alternatively, go to the retro meeting & type "lookatme". then click the button that appears.

TEST
- [x] when tab 1 hits the Add Video button, video is displayed on tab 1 & 2
- [x] when tab 2 hits refresh, they still see the video
- [x] when tab 1 toggles video off, tab 1 & 2 see the avatar instantly
- [x] when tab 2 refreshes, they see the avatar
- [x] when tab 1 toggles video back on, tab 1 & 2 see the video instantly
- [x] when tab 2 refreshes, they see the video
- [x] when tab 2 hits add video, tab 2 sees 1 local & 1 remote video
- [x] disabling video makes the webcam hardware light turn off
- [x] disabling video + audio makes the tab "recording" icon turn off

### CODE REVIEW

**Comment by mattkrick** [comment]
@jordanh ready for review.

File: src/server/utils/getPubSub.ts [hunk 1]
```diff
@@ -3,11 +3,11 @@ import {RedisPubSub} from 'graphql-redis-subscriptions'
 
 getDotenv()
 
-let pubsub
+let pubsub: RedisPubSub | undefined
 const getPubSub = () => {
   if (!pubsub) {
     pubsub = new RedisPubSub({
-      connection: process.env.REDIS_URL
+      connection: process.env.REDIS_URL as any
```
Comment (jordanh) [inline src/server/utils/getPubSub.ts]: nit: shouldn't this be `as string`?

File: src/server/wrtc/signalServer/WebSocketContext.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import ConnectionChunk from './ConnectionChunk'
+
+type WebSocketId = string
+export default class WebSocketContext {
+  userId?: string
+  createdAt: number
+  // FIXME: rename acceptedOffers since it's both offers + answers
+  acceptedOffers: {[connectionId: string]: WebSocketId} = {}
```
Comment (jordanh) [inline src/server/wrtc/signalServer/WebSocketContext.ts]: Howabout just good 'ol `connections`?

File: src/universal/utils/swarm/MediaSwarm.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+      video.srcObject = new MediaStream([track])
+      video.onloadedmetadata = () => {
+        console.log('METADATA LOADED (chrome unmuted workaround)')
+        if (track.readyState === 'live') {
+          setStreamState({[field]: true})
+        }
+      }
+
+      track.onended = () => {
+        console.log('ENDED')
```
Comment (jordanh) [inline src/universal/utils/swarm/MediaSwarm.ts]: You want to keep these console log events?

File: webpack/webpack.config.dll.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ module.exports = {
       '@babel/polyfill',
       // '@clusterws/cws',
       // '@google-cloud/language',
-      '@mattkrick/fast-rtc-peer',
+      // '@mattkrick/fast-rtc-peer', can put back in after RTC branch is merged
```
Comment (jordanh) [inline webpack/webpack.config.dll.js]: It's now 100% `fast-rtc-swarm` tho, right?

File: src/universal/utils/swarm/MediaSwarm.ts [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+export default class MediaSwarm extends FastRTCSwarm {
+  trebuchet: Trebuchet
+  dispatchState: Dispatch<ReducerAction<typeof reducerSwarm>>
+  userId: string
+  localStreams = {
+    cam: {...qualityDict},
+    screen: {...qualityDict}
+  }
+
+  constructor (config: Config) {
```
Comment (jordanh) [inline src/universal/utils/swarm/MediaSwarm.ts]: Right now, everybody is 64 x 64. In the (near) future, I imagine the person with focus will be some multiple higher than that. How will it work? Will one peer send an event to another saying "please give me your hq" stream?

**Comment by jordanh** [comment]
# Test Cycle 1

## Tests
- [ ] when tab 1 hits the Add Video button, video is displayed on tab 1 & 2
- [ ] when tab 2 hits refresh, they still see the video
- [ ] when tab 1 toggles video off, tab 1 & 2 see the avatar instantly
- [ ] when tab 2 refreshes, they see the avatar
- [ ] when tab 1 toggles video back on, tab 1 & 2 see the video instantly
- [ ] when tab 2 refreshes, they see the video
- [ ] when tab 2 hits add video, tab 2 sees 1 local & 1 remote video
- [ ] disabling video makes the webcam hardware light turn off
- [ ] disabling video + audio makes the tab "recording" icon turn off

**Comment by jordanh** [comment]
@mattkrick first test cycle done! Not sure what's the dilly. I think I got all dependences and such...

File: src/server/utils/getPubSub.ts (hunk 1 above)
Comment (mattkrick) [reply src/server/utils/getPubSub.ts]: nah, the typings expect an object, but we're using the shorthand which is a string

File: webpack/webpack.config.dll.js (hunk 4 above)
Comment (mattkrick) [reply webpack/webpack.config.dll.js]: yeah, i can probably remove peer or at least bump it to a dev dependency

File: src/universal/utils/swarm/MediaSwarm.ts (hunk 5 above)
Comment (mattkrick) [reply src/universal/utils/swarm/MediaSwarm.ts]: exactly right! a user creates up to 3 streams and serves up the appropriate stream to the peer.

File: src/universal/utils/swarm/MediaSwarm.ts (hunk 3 above)
Comment (mattkrick) [reply src/universal/utils/swarm/MediaSwarm.ts]: for alpha definitely. it'll help with sentry errors & debugging

**Comment by mattkrick** [comment]
@jordanh did you get any server errors? the graphql server guarantees that `featureFlags` exists, so i'm stumped.

**Comment by jordanh** [comment]
# Test Cycle 2

## Tests
- [ ] when tab 1 hits the Add Video button, video is displayed on tab 1 & 2
- [ ] when tab 2 hits refresh, they still see the video
- [ ] when tab 1 toggles video off, tab 1 & 2 see the avatar instantly
- [ ] when tab 2 refreshes, they see the avatar
- [ ] when tab 1 toggles video back on, tab 1 & 2 see the video instantly
- [ ] when tab 2 refreshes, they see the video
- [ ] when tab 2 hits add video, tab 2 sees 1 local & 1 remote video
- [ ] disabling video makes the webcam hardware light turn off
- [ ] disabling video + audio makes the tab "recording" icon turn off

## Test Notes

**Comment by jordanh** [comment]
@mattkrick Test Cycle 2 complete – I seeing some TypeScript warnings that should probably get cleaned up...

**Comment by mattkrick** [comment]
huh, i didn't see those errors, but i'm running ts 3.3.3333. they'll probably get cleaned up by a new version of ts

**Pull request merged**